README.md
# Time Ago In Words

**Localizable relative time resolver**

[![Build Status](https://travis-ci.org/bluemarblepayroll/time-ago-in-words.svg?branch=master)](https://travis-ci.org/bluemarblepayroll/time-ago-in-words)

This library lets you easily convert a JavaScript Date object to a relative time span description.  For example, let's say the
current date and time is January 1, 2018 12:00:00 PM, then January 10, 2018 12:00:00 PM would resolve to: *in 9 days*.  More general examples:

| Date/Time                       | Relative Description |
| ------------------------------- | -------------------- |
| January 1, 2016 12:00:00 PM)    | 2 years ago          |
| November 1, 2018 12:00:00 PM)   | 2 months ago         |
| December 30, 2017 12:00:00 PM)  | 2 days ago           |
| January 1, 2018 4:00:00 AM)     | 8 hours ago          |
| January 1, 2018 11:00:00 AM)    | 1 hour ago           |
| January 1, 2018 11:50:00 AM)    | 10 minutes ago       |
| January 1, 2018 11:58:00 AM)    | 2 minutes ago        |
| January 1, 2018 11:59:01 AM)    | a minute ago         |
| January 1, 2018 12:00:01 PM)    | in a minute          |
| January 1, 2018 12:02:00 PM)    | in 2 minutes         |
| January 1, 2018 12:10:00 PM)    | in 10 minutes        |
| January 1, 2018 1:00:00 PM)     | in 1 hour            |
| January 1, 2018 8:00:00 PM)     | in 8 hours           |
| January 3, 2018 12:00:00 PM)    | in 2 days            |
| April 1, 2018 12:00:00 PM)      | in 2 months          |
| January 1, 2020 12:00:00 PM)    | in 2 years           |

Another important feature of this library is its ability to localize any of the relative expressions.  See examples below for more information.

## Credit

Rail's time_ago_in_words along with (this gist)[https://gist.github.com/boxnos/5896114] helped inspire this library's API and internal implementation.

## Installation

This library could be consumed as either a pure TypeScript library or as its trans-compiled ES2015 JavaScript counterpart.

To install through NPM:

````
npm install --save @bluemarblepayroll/time-ago-in-words
````

To install through Yarn:

````
yarn add @bluemarblepayroll/time-ago-in-words
````

## Examples

### A Basic Example

Suppose you just want to get the relative time between a Date and now (in our case lets use January 1, 2018 12:00:00 PM):

````
import { timeAgoInWords } from '@bluemarblepayroll/time-ago-in-words';

let from = new Date(2017, 11, 31, 23); # December 31, 2017 11:00:00 PM

let value = timeAgoInWords(from); # "13 hours ago"
````

It also works for future times as well:

````
let from = new Date(2020, 0, 1, 12); # January 1, 2020 12:00:00 PM

let value = timeAgoInWords(from); # "in 2 years"
````

### Explicit Time Span Example

You can also pass in each part of the time span (from and to).  Using the above examples, respectively, would become:

````
let from = new Date(2017, 11, 31, 23); # December 31, 2017 11:00:00 PM
let to = new Date(2018, 0, 1, 12); # January 1, 2018 12:00:00 PM

let value = timeAgoInWords(from, to); # "13 hours ago"
````

````
let from = new Date(2020, 0, 1, 12); # January 1, 2020 12:00:00 PM
let to = new Date(2018, 0, 1, 12); # January 1, 2018 12:00:00 PM

let value = timeAgoInWords(from, to); # "in 2 years"
````

### Custom Localization Example

It is also possible to define custom translations.  For all possible translations, see baseTranslations assignment within the file: ./lib/i18n.ts.

Based on our previous examples, let's say we wanted to localize the following phrases:

* "%s hours ago" => "hace %s horas"
* "in %s years" => "en %s años"

Based on our initial examples, respectively, we can achieve this like so:

````
import { timeAgoSet } from '@bluemarblepayroll/time-ago-in-words';

timeAgoSet({ '%s hours ago': 'hace %s horas' });

let from = new Date(2017, 11, 31, 23); # December 31, 2017 11:00:00 PM

let value = timeAgoInWords(from); # "hace 13 horas"
````

````
import { timeAgoSet } from '@bluemarblepayroll/time-ago-in-words';

timeAgoSet({ 'in %s years': 'en %s años' });

let from = new Date(2020, 0, 1, 12); # January 1, 2020 12:00:00 PM

let value = timeAgoInWords(from); # "en 2 años"
````

If you wish to revert back to default translations:

````
import { timeAgoReset } from '@bluemarblepayroll/time-ago-in-words';

timeAgoReset();
````

## Contributing

### Development Environment Configuration

Basic steps to take to get this repository compiling:

1. Install [Node.js](https://nodejs.org) (check package.json for versions supported.)
2. Install Yarn package manager (npm install -g yarn)
3. Clone the repository (git clone git@github.com:bluemarblepayroll/time-ago-in-words.git)
4. Navigate to the root folder (cd time-ago-in-words)
5. Install dependencies (yarn)

### Compiling

To compile the TypeScript source down to native JavaScript, run:

````
yarn run build
````

### Running Tests

To execute the test suite run:

````
yarn run test
````

### Linting

````
yarn run lint
````

### Publishing

Note: ensure you have proper authorization before trying to publish new versions.

After code changes have successfully gone through the Pull Request review process then the following steps should be followed for publishing new versions:

1. Merge Pull Request into master
2. Update package.json [version number](https://semver.org/)
3. Update CHANGELOG.md
4. Push master to remote and ensure CI builds master successfully
5. Build the project locally: `yarn run build`
6. Perform a dry run: `npm publish --access public --dry-run`.  Inspect packaging, ensure all files (including dist) are included.
7. Publish package to NPM: `npm publish --access public`
8. Tag master with new version: `git tag <version>`
9. Push tags remotely: `git push origin --tags`

## License

This project is MIT Licensed.


CHANGELOG.md
## 2.0.0 (November 1, 2018)

* Breaking Change: Removed Processor namespace from main interface.  You have to now directly import required methods.
  - Good: `import * as TimeAgo from '@bluemarblepayroll/time-ago-in-words';`
  - Good: `import { timeAgoReset, timeAgoSet, timeAgoInWords } from '@bluemarblepayroll/time-ago-in-words';`
  - Bad: `import { TimeAgo } from '@bluemarblepayroll/time-ago-in-words';`
* Breaking Change: Renamed methods.
  - TimeAgo#inWords => timeAgoInWords
  - TimeAgo#reset => timeAgoReset
  - TimeAgo#set => timeAgoSet


test/time_ago.spec.ts
/**
 * Copyright (c) 2018-present, Blue Marble Payroll, LLC
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { expect } from "chai";
import { timeAgoInWords, timeAgoReset, timeAgoSet } from "../lib/time_ago";

describe("when using custom translations", () => {

  beforeEach( () => {
    timeAgoReset();
  });

  it("should correctly translate: less than a minute ago", () => {

    timeAgoSet({ "less than a minute ago": "abc" });

    const from = new Date(2017, 11, 31, 23, 59, 1);
    const to = new Date(2018, 0);

    const result = timeAgoInWords(from, to);
    expect(result).to.equal("abc");
  });

  it("should correctly translate: in a minute", () => {

    timeAgoSet({ "in a minute": "abc" });

    const from = new Date(2018, 0, 1, 0, 0, 1);
    const to = new Date(2018, 0);

    const result = timeAgoInWords(from, to);
    expect(result).to.equal("abc");
  });

});

describe("TimeAgo#inWords", () => {

  beforeEach( () => {
    timeAgoReset();
  });

  it("when from date is more than 2 years back in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(1805, 0),  to, "213 years ago" ],
      [ new Date(1905, 0),  to, "113 years ago" ],
      [ new Date(1975, 0),  to, "43 years ago" ],
      [ new Date(1995, 0),  to, "23 years ago" ],
      [ new Date(2000, 0),  to, "18 years ago" ],
      [ new Date(2005, 0),  to, "13 years ago" ],
      [ new Date(2010, 0),  to, "8 years ago" ],
      [ new Date(2015, 0),  to, "3 years ago" ],
      [ new Date(2016, 0),  to, "2 years ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when from date is more than 1 but less than 2 years back in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2016, 1),  to, "1 year ago" ],
      [ new Date(2016, 2),  to, "1 year ago" ],
      [ new Date(2016, 3),  to, "1 year ago" ],
      [ new Date(2016, 4),  to, "1 year ago" ],
      [ new Date(2016, 5),  to, "1 year ago" ],
      [ new Date(2016, 6),  to, "1 year ago" ],
      [ new Date(2016, 7),  to, "1 year ago" ],
      [ new Date(2016, 8),  to, "1 year ago" ],
      [ new Date(2016, 9),  to, "1 year ago" ],
      [ new Date(2016, 10), to, "1 year ago" ],
      [ new Date(2016, 11), to, "1 year ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when from date is less than 1 year back in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2017, 0),      to, "12 months ago" ],
      [ new Date(2017, 1),      to, "11 months ago" ],
      [ new Date(2017, 2),      to, "10 months ago" ],
      [ new Date(2017, 3),      to, "9 months ago" ],
      [ new Date(2017, 4),      to, "8 months ago" ],
      [ new Date(2017, 5),      to, "7 months ago" ],
      [ new Date(2017, 6),      to, "6 months ago" ],
      [ new Date(2017, 7),      to, "5 months ago" ],
      [ new Date(2017, 8),      to, "4 months ago" ],
      [ new Date(2017, 9),      to, "3 months ago" ],
      [ new Date(2017, 10),     to, "2 months ago" ],
      [ new Date(2017, 11),     to, "1 month ago" ],
      [ new Date(2017, 11, 0),  to, "1 month ago" ],
      [ new Date(2017, 11, 1),  to, "1 month ago" ],
      [ new Date(2017, 11, 2),  to, "1 month ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when from date is less than 1 month back in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2017, 11, 3),      to, "29 days ago" ],
      [ new Date(2017, 11, 6),      to, "26 days ago" ],
      [ new Date(2017, 11, 9),      to, "23 days ago" ],
      [ new Date(2017, 11, 12),     to, "20 days ago" ],
      [ new Date(2017, 11, 15),     to, "17 days ago" ],
      [ new Date(2017, 11, 18),     to, "14 days ago" ],
      [ new Date(2017, 11, 21),     to, "11 days ago" ],
      [ new Date(2017, 11, 24),     to, "8 days ago" ],
      [ new Date(2017, 11, 27),     to, "5 days ago" ],
      [ new Date(2017, 11, 28),     to, "4 days ago" ],
      [ new Date(2017, 11, 29),     to, "3 days ago" ],
      [ new Date(2017, 11, 30),     to, "2 days ago" ],
      [ new Date(2017, 11, 31),     to, "1 day ago" ],
      [ new Date(2017, 11, 31, 0),  to, "1 day ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when from date is less than 1 day back up until 15 minutes in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2017, 11, 31, 1),      to, "23 hours ago" ],
      [ new Date(2017, 11, 31, 4),      to, "20 hours ago" ],
      [ new Date(2017, 11, 31, 7),      to, "17 hours ago" ],
      [ new Date(2017, 11, 31, 10),     to, "14 hours ago" ],
      [ new Date(2017, 11, 31, 13),     to, "11 hours ago" ],
      [ new Date(2017, 11, 31, 16),     to, "8 hours ago" ],
      [ new Date(2017, 11, 31, 19),     to, "5 hours ago" ],
      [ new Date(2017, 11, 31, 22),     to, "2 hours ago" ],
      [ new Date(2017, 11, 31, 23),     to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 0),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 1),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 2),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 3),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 4),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 5),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 6),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 7),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 8),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 9),  to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 10), to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 11), to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 12), to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 13), to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 14), to, "1 hour ago" ],
      [ new Date(2017, 11, 31, 23, 15), to, "1 hour ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when from date is less than 45 minutes in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2017, 11, 31, 23, 16), to, "44 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 17), to, "43 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 18), to, "42 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 19), to, "41 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 20), to, "40 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 25), to, "35 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 30), to, "30 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 40), to, "20 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 50), to, "10 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 58), to, "2 minutes ago" ],
      [ new Date(2017, 11, 31, 23, 59), to, "1 minute ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is less an 1 minute in the past", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2017, 11, 31, 23, 59, 1),  to, "less than a minute ago" ],
      [ new Date(2017, 11, 31, 23, 59, 10), to, "less than a minute ago" ],
      [ new Date(2017, 11, 31, 23, 59, 20), to, "less than a minute ago" ],
      [ new Date(2017, 11, 31, 23, 59, 30), to, "less than a minute ago" ],
      [ new Date(2017, 11, 31, 23, 59, 40), to, "less than a minute ago" ],
      [ new Date(2017, 11, 31, 23, 59, 50), to, "less than a minute ago" ],
      [ new Date(2017, 11, 31, 23, 59, 58), to, "less than a minute ago" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is less than 1 minute in the future", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2018, 0, 1, 0, 0, 1),  to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 2),  to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 10), to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 20), to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 30), to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 40), to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 50), to, "in a minute" ],
      [ new Date(2018, 0, 1, 0, 0, 59), to, "in a minute" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is more than 1 minute in the future", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2018, 0, 1, 0, 2),   to, "in 2 minutes" ],
      [ new Date(2018, 0, 1, 0, 3),   to, "in 3 minutes" ],
      [ new Date(2018, 0, 1, 0, 4),   to, "in 4 minutes" ],
      [ new Date(2018, 0, 1, 0, 5),   to, "in 5 minutes" ],
      [ new Date(2018, 0, 1, 0, 10),  to, "in 10 minutes" ],
      [ new Date(2018, 0, 1, 0, 20),  to, "in 20 minutes" ],
      [ new Date(2018, 0, 1, 0, 30),  to, "in 30 minutes" ],
      [ new Date(2018, 0, 1, 0, 40),  to, "in 40 minutes" ],
      [ new Date(2018, 0, 1, 0, 40),  to, "in 40 minutes" ],
      [ new Date(2018, 0, 1, 0, 41),  to, "in 41 minutes" ],
      [ new Date(2018, 0, 1, 0, 42),  to, "in 42 minutes" ],
      [ new Date(2018, 0, 1, 0, 43),  to, "in 43 minutes" ],
      [ new Date(2018, 0, 1, 0, 44),  to, "in 44 minutes" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is more than 45 minutes and less than 1 day in the future", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2018, 0, 1, 0, 45),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 46),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 47),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 48),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 49),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 50),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 55),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 0, 59),   to, "in 1 hour" ],
      [ new Date(2018, 0, 1, 1, 0),    to, "in 1 hour" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is more than 1 hour and less than 1 day in the future", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2018, 0, 1, 2),   to, "in 2 hours" ],
      [ new Date(2018, 0, 1, 3),   to, "in 3 hours" ],
      [ new Date(2018, 0, 1, 4),   to, "in 4 hours" ],
      [ new Date(2018, 0, 1, 5),   to, "in 5 hours" ],
      [ new Date(2018, 0, 1, 6),   to, "in 6 hours" ],
      [ new Date(2018, 0, 1, 7),   to, "in 7 hours" ],
      [ new Date(2018, 0, 1, 8),   to, "in 8 hours" ],
      [ new Date(2018, 0, 1, 9),   to, "in 9 hours" ],
      [ new Date(2018, 0, 1, 10),  to, "in 10 hours" ],
      [ new Date(2018, 0, 1, 20),  to, "in 20 hours" ],
      [ new Date(2018, 0, 1, 23),  to, "in 23 hours" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is 1 or more months and less than 1 year in the future", () => {
    const to = new Date(2018, 0);

    const tests = [
      [ new Date(2018, 1),  to, "in 1 month" ],
      [ new Date(2018, 2),  to, "in 2 months" ],
      [ new Date(2018, 3),  to, "in 3 months" ],
      [ new Date(2018, 4),  to, "in 4 months" ],
      [ new Date(2018, 5),  to, "in 5 months" ],
      [ new Date(2018, 10), to, "in 10 months" ],
      [ new Date(2018, 11), to, "in 11 months" ],
    ];

    tests.forEach((x) => {
      const result = timeAgoInWords(x[0] as Date, x[1] as Date);
      expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
    });
  });

  it("when the from date is 1 or more years in the future", () => {
      const to = new Date(2018, 0);

      const tests = [
        [ new Date(2019, 0),  to, "in 1 year" ],
        [ new Date(2020, 0),  to, "in 2 years" ],
        [ new Date(2021, 0),  to, "in 3 years" ],
        [ new Date(2022, 0),  to, "in 4 years" ],
        [ new Date(2023, 0),  to, "in 5 years" ],
        [ new Date(2033, 0),  to, "in 15 years" ],
        [ new Date(2133, 0),  to, "in 115 years" ],
        [ new Date(2233, 0),  to, "in 215 years" ],
      ];

      tests.forEach((x) => {
        const result = timeAgoInWords(x[0] as Date, x[1] as Date);
        expect(result).to.equal(x[2], `Failed on: ${x[0]}`);
      });
  });
});


lib/i18n.ts
/**
 * Copyright (c) 2018-present, Blue Marble Payroll, LLC
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const baseTranslations: Record<string, string> = {
  "in %s year": null,
  "in %s years": null,
  // We want to disable linting this list of keys.
  // It makes more sense sorting them 'chronologically'.
  // tslint:disable-next-line:object-literal-sort-keys
  "in %s month": null,
  "in %s months": null,
  "in %s day": null,
  "in %s days": null,
  "in %s hour": null,
  "in %s hours": null,
  "in %s minutes": null,
  "in a minute": null,
  "less than a minute ago": null,
  "%s minute ago": null,
  "%s minutes ago": null,
  "%s hour ago": null,
  "%s hours ago": null,
  "%s day ago": null,
  "%s days ago": null,
  "%s month ago": null,
  "%s months ago": null,
  "%s year ago": null,
  "%s years ago": null,
};

let translations: Record<string, string>;

export function reset(): void {
  translations = { };

  for (const key of Object.keys(baseTranslations)) {
    translations[key] = baseTranslations[key];
  }
}

reset();

export function set(newTranslations: Record<string, string>): void {
  if (!newTranslations) {
    return;
  }

  for (const key of Object.keys(newTranslations)) {
    translations[key] = newTranslations[key];
  }
}

function get(phrase: string): string {
  if (!translations[phrase]) {
    return phrase || "";
  }

  return translations[phrase] || "";
}

export function translate(phrase: string, value: string): string {
  const translatedPhrase: string = get(phrase);

  return translatedPhrase.replace("%s", value);
}


lib/binary_search.ts
/**
 * Copyright (c) 2018-present, Blue Marble Payroll, LLC
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export interface IEntry {
  key: number;
}

function innerSearch(array: IEntry[], value: number, lower: number, pos: number, upper: number): IEntry {
  if (array[pos].key <= value && value < array[pos + 1].key) {
    return array[pos];
  } else if (value < array[pos].key) {
    return innerSearch(array, value, lower, Math.floor((lower + pos - 1) / 2), pos - 1);
  }

  return innerSearch(array, value, pos + 1, Math.floor((pos + 1 + upper) / 2), upper);
}

export function binarySearch(array: IEntry[], value: number): IEntry {
  return innerSearch(array, value, 0, Math.floor((array.length - 1) / 2), array.length - 1);
}


lib/time_ago.ts
/**
 * Copyright (c) 2018-present, Blue Marble Payroll, LLC
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { binarySearch, IEntry } from "./binary_search";
import { reset, set, translate } from "./i18n";

enum Unit {
  Years = "year",
  Months = "month",
  Days = "day",
  Hours = "hour",
  Minutes = "minute",
}

enum Tense {
  Past,
  Future,
}

class TimeSpan {
  public unit: Unit;
  public tense: Tense;
  public value: number;

  constructor(unit: Unit, tense: Tense, value: number) {
    this.unit = unit;
    this.tense = tense;
    this.value = value;
  }

  public template(): string {
    if (this.unit === Unit.Minutes && this.value === 0 && this.tense === Tense.Future) {
      return `in a minute`;
    } else if (this.unit === Unit.Minutes && this.value === 0 && this.tense === Tense.Past) {
      return `less than a minute ago`;
    } else if (this.tense === Tense.Past) {
      return `%s ${this.noun()} ago`;
    } else if (this.tense === Tense.Future) {
      return `in %s ${this.noun()}`;
    } else {
      throw new Error(`Tense cannot be handled: ${this.tense}`);
    }
  }

  private noun(): string {
    if (this.value > 1) {
      return `${this.unit}s`;
    }

    return this.unit;
  }
}

type ValueResolver = (minutes: number) => number;

interface IMinutesEntry extends IEntry {
  unit: Unit;
  tense: Tense;
  valueResolver: ValueResolver;
}

const minuteRanges: IMinutesEntry[] = [
  {
    key: Number.NEGATIVE_INFINITY,
    tense: Tense.Future,
    unit: Unit.Years,
    valueResolver: (m) => parseInt(Math.abs(m / 525960).toFixed(), 10),
  },
  {
    key: -529600,
    tense: Tense.Future,
    unit: Unit.Years,
    valueResolver: (_) => 1,
  },
  {
    key: -486400,
    tense: Tense.Future,
    unit: Unit.Months,
    valueResolver: (m) => parseInt(Math.abs(m / 43200).toFixed(), 10),
  },
  {
    key: -43200,
    tense: Tense.Future,
    unit: Unit.Months,
    valueResolver: (_) => 1,
  },
  {
    key: -41760,
    tense: Tense.Future,
    unit: Unit.Days,
    valueResolver: (m) => parseInt(Math.abs(m / 1440).toFixed(), 10),
  },
  {
    key: -1440,
    tense: Tense.Future,
    unit: Unit.Days,
    valueResolver: (_) => 1,
  },
  {
    key: -1380,
    tense: Tense.Future,
    unit: Unit.Hours,
    valueResolver: (m) => parseInt(Math.abs(m / 60).toFixed(), 10),
  },
  {
    key: -45,
    tense: Tense.Future,
    unit: Unit.Hours,
    valueResolver: (_) => 1 },
  {
    key: -44,
    tense: Tense.Future,
    unit: Unit.Minutes,
    valueResolver: (m) => parseInt(Math.abs(m).toFixed(), 10),
  },
  {
    key: -1,
    tense: Tense.Future,
    unit: Unit.Minutes,
    valueResolver: (_) => 0,
  },
  {
    key: 0,
    tense: Tense.Past,
    unit: Unit.Minutes,
    valueResolver: (_) => 0,
  },
  {
    key: 1,
    tense: Tense.Past,
    unit: Unit.Minutes,
    valueResolver: (_) => 1,
  },
  {
    key: 2,
    tense: Tense.Past,
    unit: Unit.Minutes,
    valueResolver: (m) => parseInt(m.toFixed(), 10),
  },
  {
    key: 45,
    tense: Tense.Past,
    unit: Unit.Hours,
    valueResolver: (_) => 1,
  },
  {
    key: 90,
    tense: Tense.Past,
    unit: Unit.Hours,
    valueResolver: (m) => parseInt((m / 60).toFixed(), 10),
  },
  {

    key: 1440,
    tense: Tense.Past,
    unit: Unit.Days,
    valueResolver: (_) => 1,
  },
  {
    key: 2880,
    tense: Tense.Past,
    unit: Unit.Days,
    valueResolver: (m) => parseInt((m / 1440).toFixed(), 10),
  },
  {
    key: 43200,
    tense: Tense.Past,
    unit: Unit.Months,
    valueResolver: (_) => 1,
  },
  {
    key: 86400,
    tense: Tense.Past,
    unit: Unit.Months,
    valueResolver: (m) => parseInt((m / 43200).toFixed(), 10),
  },
  {
    key: 529600,
    tense: Tense.Past,
    unit: Unit.Years,
    valueResolver: (_) => 1,
  },
  {
    key: 1051200,
    tense: Tense.Past,
    unit: Unit.Years,
    valueResolver: (m) => parseInt((m / 525960).toFixed(), 10),
  },
  {
    key: Number.MAX_VALUE,
    tense: null,
    unit: null,
    valueResolver: (_) => null,
  },
];

function getTimeSpan(from: Date, to: Date): TimeSpan {
  const minutes: number = (+to - +from) / 60000;

  const match: IMinutesEntry = binarySearch(minuteRanges, minutes) as IMinutesEntry;

  return new TimeSpan(match.unit, match.tense, match.valueResolver(minutes));
}

export function timeAgoInWords(from: Date, to: Date = new Date()): string {
  const timeSpan: TimeSpan = getTimeSpan(from, to);

  return translate(timeSpan.template(), timeSpan.value.toString());
}

export {
  reset as timeAgoReset,
  set as timeAgoSet,
};
